circuits/cpp/src/aztec3/circuits/kernel/private/native_private_kernel_circuit_init.cpp

```diff
@@ -0,0 +1,195 @@
+                       "Users cannot make a delegatecall",
+                       CircuitErrorCode::PRIVATE_KERNEL__UNSUPPORTED_OP);
+    composer.do_assert(this_call_stack_item.public_inputs.call_context.is_static_call == false,
+                       "Users cannot make a static call",
+                       CircuitErrorCode::PRIVATE_KERNEL__UNSUPPORTED_OP);
+
+    // The below also prevents delegatecall/staticcall in the base case
+    composer.do_assert(this_call_stack_item.public_inputs.call_context.storage_contract_address ==
+                           this_call_stack_item.contract_address,
+                       "Storage contract address must be that of the called contract",
+                       CircuitErrorCode::PRIVATE_KERNEL__CONTRACT_ADDRESS_MISMATCH);
+
+    // TODO: Assert that the previous kernel data is empty. (Or rather, the verify_proof() function needs a valid
+    // dummy proof and vk to complete execution, so actually what we want is for that mockvk to be
+    // hard-coded into the circuit and assert that that is the one which has been used in the base case).
+}
+
+void update_end_values(PrivateKernelInputsInit<NT> const& private_inputs, KernelCircuitPublicInputs<NT>& public_inputs)
+{
+    // We only initialzed constants member of public_inputs so far. Therefore, there must not be any
+    // new nullifiers as part of public_inputs.
+    ASSERT(is_array_empty(public_inputs.end.new_nullifiers));
+
+    // Since it's the first iteration, we need to push the the tx hash nullifier into the `new_nullifiers` array
+    array_push(public_inputs.end.new_nullifiers, private_inputs.signed_tx_request.hash());
+
+    // Nonce nullifier
+    // DANGER: This is terrible. This should not be part of the protocol. This is an intentional bodge to reach a
+    // milestone. This must not be the way we derive nonce nullifiers in production. It can be front-run by other
+    // users. It is not domain separated. Naughty.
+    array_push(public_inputs.end.new_nullifiers, private_inputs.signed_tx_request.tx_request.nonce);
+}
+
+
+// NOTE: THIS IS A VERY UNFINISHED WORK IN PROGRESS.
+// TODO: decide what to return.
+// TODO: is there a way to identify whether an input has not been used by ths circuit? This would help us more-safely
+// ensure we're constraining everything.
+KernelCircuitPublicInputs<NT> native_private_kernel_circuit_initial(DummyComposer& composer,
+                                                                    PrivateKernelInputsInit<NT> const& private_inputs)
+{
+    // We'll be pushing data to this during execution of this circuit.
+    KernelCircuitPublicInputs<NT> public_inputs{};
+
+    // Do this before any functions can modify the inputs.
+    initialise_end_values(private_inputs, public_inputs);
+
+    validate_inputs(composer, private_inputs);
+
+    // validate_this_private_call_against_tx_request(composer, private_inputs);
+
+    // common_validate_call_stack(composer, private_inputs.private_call);
```

Comment:
Add back rahul's scary todo here with links to tickets:

* https://github.com/AztecProtocol/aztec-packages/issues/499
* https://github.com/AztecProtocol/aztec-packages/issues/671
* https://github.com/AztecProtocol/aztec-packages/issues/672

Reply:
Added by Jean.